### PR TITLE
recognize assigning .init of all zeros, replace with memset

### PR DESCRIPTION
**Pull request by WalterBright**
Structs are often initialized or assigned values of all 0 bits. This usually results in copying the `.init` global into the struct. This PR recognizes some of those cases and replaces them with `memset`'ing them to zero. This is faster, and a step towards eliminating the `.init` global altogether.

### CODE REVIEW

**Comment by WalterBright** [comment]
Related to https://issues.dlang.org/show_bug.cgi?id=14639

File: src/dmd/e2ir.d [hunk 1]
```diff
@@ -3123,6 +3123,36 @@ elem *toElem(Expression e, IRState *irs)
...
+                            return false;
+                        }
+                        return true;
+                    }
+
+                    /* Use a memset to 0
+                     */
+                    if ((sle.useStaticInit ||
+                         sle.elements && allZeroBits(*sle.elements) && !sle.sd.isNested()) &&
+                        ae.e2.type.isZeroInit(ae.e2.loc))
```
Comment (n8sh) [inline src/dmd/e2ir.d]: Should this be
```d
(sle.useStaticInit && ae.e2.type.isZeroInit(ae.e2.loc))
||
(sle.elements && allZeroBits(*sle.elements) && !sle.sd.isNested())
```
?

Comment (WalterBright) [reply src/dmd/e2ir.d]: no because sle.elements does not include default initializers.

**Comment by WalterBright** [comment]
```
C++ interop tests / Run (macOS-10.15, clang-4.0.0) (pull_request) Failing after 46m
```
So I click for "Details", and there are no details, no log, just it failed.

**Comment by NilsLankila** [comment]
Does this include init with void gaps ?, e.g

```
struct F { int i; void* p = void; int j; }
```
because in this case the gap can be ignore and filled with 0 as well.

**Comment by WalterBright** [comment]
Looks like the C++ heisenbug disappeared.

**Review by atilaneves — APPROVED** [review]

**Comment by WalterBright** [comment]
@atilaneves wunderbar!

**Pull request merged**